control-base/src/main/java/fi/nls/oskari/control/layer/DataProviderHandler.java

```diff
@@ -17,19 +25,97 @@
 public class DataProviderHandler extends RestActionHandler {
 
 	private static final String PARAM_ID = "id";
+	private static final String PARAM_DELETE_LAYERS = "deleteLayers";
 
-	private DataProviderService service = ServiceFactory.getDataProviderService();
+	private DataProviderService dataProviderService = ServiceFactory.getDataProviderService();
+	private OskariLayerService mapLayerService = ServiceFactory.getMapLayerService();
 
 	@Override
 	public void handleGet(ActionParameters params) throws ActionException {
-		
+
 		final int id = params.getRequiredParamInt(PARAM_ID);
-		
+
 		try {
-			DataProvider d = service.find(id);
+			DataProvider d = dataProviderService.find(id);
 			ResponseHelper.writeResponse(params, d.getAsJSON());
 		} catch (Exception e) {
 			throw new ActionException("Data provider query failed", e);
 		}
 	}
+
+	@Override
+	public void handleDelete(ActionParameters params) throws ActionException {
+
+		final int id = params.getRequiredParamInt(PARAM_ID);
+		final boolean deleteLayers = params.getRequiredParamBoolean(PARAM_DELETE_LAYERS);
+
+		if (!dataProviderService.hasPermissionToUpdate(params.getUser(), id)) {
+			throw new ActionDeniedException("Unauthorized user tried to remove data provider id=" + id);
+		}
+
+		if (deleteLayers) {
+			deleteDataProviderAndLayers(id, params);
+		} else {
+			deleteDataProvider(id, params);
+		}
```

Comment:
It would be easier to instantly see whats going on if there was a method deleteLayersForProvider()  and removeProviderFromLayers() which would be called based on the boolean. After the conditional statements you could call deleteDataProvider() that would return the removed provider. Then you could write the response based on that in handleDelete.

Reply:
Fixed but little bit different way than suggested.